src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -364,38 +365,38 @@ private function getTagVersions(array $tagsByKey, bool $persistTags): array
         foreach ($tagVersions as $tag => $version) {
             $tags[$tag.static::TAGS_PREFIX] = $tag;
             $knownTagVersion = $this->knownTagVersions[$tag] ?? [0, null];
-            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now - $knownTagVersion[0] >= $this->knownTagVersionsTtl) {
-                // reuse previously fetched tag versions up to the ttl
+            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now > $knownTagVersion[0]) {
+                // reuse previously fetched tag versions until the expiration
+                unset($this->knownTagVersions[$tag];
                 $fetchTagVersions = true;
             }
-            unset($this->knownTagVersions[$tag]); // For LRU tracking
-            if ([0, null] !== $knownTagVersion) {
-                $this->knownTagVersions[$tag] = $knownTagVersion;
-            }
         }
 
         if (!$fetchTagVersions) {
             return $tagVersions;
         }
 
+        if (\count($this->knownTagVersions) > self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS && $now > reset($this->knownTagVersions)[0]) {
+            array_splice($this->knownTagVersions, 0, \count($this->knownTagVersions) >> 2);
+        }
+
         $newTags = [];
         $newVersion = null;
+        $expiration = $now + $this->knownTagVersionsTtl;
         foreach ($this->tags->getItems(array_keys($tags)) as $tag => $version) {
-            if (!$version->isHit()) {
+            unset($this->knownTagVersions[$tag = $tags[$tag]]); // update FIFO
+            if ($persistTags && !$version->isHit()) {
                 $newTags[$tag] = $version->set($newVersion ??= random_bytes(6));
             }
-            $tagVersions[$tag = $tags[$tag]] = $version->get();
-            $this->knownTagVersions[$tag] = [$now, $tagVersions[$tag]];
+            if ($tagVersions[$tag] = $version->get()) {
```

Comment:
```php
if ($version->isHit()) {
    $tagVersions[$tag] = $version->get();
    $this->knownTagVersions[$tag] = [$expiration, $tagVersions[$tag]];
} elseif ($persistTags) {
    $newTags[$tag] = $version->set($newVersion ??= random_bytes(6));
}
```

Reply:
This part of code can be rewritten as follows.
```php
        $tagVersions[$tag] = null; // we must always return values for all requested tags
        if ($version->isHit()) {
            $tagVersions[$tag] = $version->get();
            $this->knownTagVersions[$tag] = [$expiration, $tagVersions[$tag]];
        } elseif ($persistTags) {
            $newTags[$tag] = $version->set($newVersion ??= random_bytes(6));
            $tagVersions[$tag] = $newVersion; // for attaching to just computed item(s)
            $this->knownTagVersions[$tag] = [$expiration, $tagVersions[$tag]]; // this is optional but may save us a round-trip
//        } else {
//            $tagVersions[$tag] = null; // the alternative to the first line
        }
```

Reply:
indeed, then we could go with
```php
if ($tagVersions[$tag] = $version->get()) {
    $this->knownTagVersions[$tag] = [$expiration, $tagVersions[$tag]];
} elseif ($persistTags) {
    $newTags[$tag] = $version->set($newVersion ??= random_bytes(6));
}```